File: rle_decompress.js
// rle_decompress.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const process = require('process');

const COLORS = {
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    reset: '\x1b[0m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

const ESCAPE = '\\';

function decompress(text) {
    if (!text) return '';
    const result = [];
    let i = 0, n = text.length;
    while (i < n) {
        const ch = text[i];
        if (ch === ESCAPE) {
            if (i + 1 < n && text[i + 1] === ESCAPE) {
                result.push(ESCAPE);
                i += 2;
                continue;
            }
            if (i + 1 >= n) throw new Error('Unexpected end after escape');
            const repeatChar = text[i + 1];
            i += 2;
            let numStr = '';
            while (i < n && text[i] >= '0' && text[i] <= '9') {
                numStr += text[i];
                i++;
            }
            if (!numStr) throw new Error('Missing number after escape');
            const count = parseInt(numStr, 10);
            for (let k = 0; k < count; k++) result.push(repeatChar);
        } else {
            result.push(ch);
            i++;
        }
    }
    return result.join('');
}

function readInput(filename) {
    if (!filename || filename === '-') {
        return fs.readFileSync(0, 'utf-8');
    }
    return fs.readFileSync(filename, 'utf-8');
}

function writeOutput(filename, content) {
    if (!filename || filename === '-') {
        process.stdout.write(content);
    } else {
        fs.writeFileSync(filename, content, 'utf-8');
    }
}

function main() {
    const args = process.argv.slice(2);
    let inputFile = null, outputFile = null, verbose = false;
    for (let i = 0; i < args.length; i++) {
        if (args[i] === '-v' || args[i] === '--verbose') {
            verbose = true;
        } else if (!inputFile) {
            inputFile = args[i];
        } else {
            outputFile = args[i];
        }
    }
    if (!inputFile) {
        console.log(colorize('Usage: node rle_decompress.js <input> [output] [-v]', 'yellow'));
        process.exit(1);
    }

    let data;
    try {
        data = readInput(inputFile);
    } catch (err) {
        console.log(colorize('Error reading input: ' + err.message, 'red'));
        process.exit(1);
    }
    const inputSize = Buffer.byteLength(data, 'utf8');

    let result;
    try {
        result = decompress(data);
    } catch (err) {
        console.log(colorize('Decompression error: ' + err.message, 'red'));
        process.exit(1);
    }
    const outputSize = Buffer.byteLength(result, 'utf8');

    if (verbose) {
        const ratio = inputSize ? outputSize / inputSize : 1;
        console.log(colorize(`Compressed size: ${inputSize} bytes`, 'yellow'));
        console.log(colorize(`Decompressed size: ${outputSize} bytes`, 'yellow'));
        console.log(colorize(`Expansion ratio: ${ratio.toFixed(2)}x`, 'green'));
    }

    try {
        writeOutput(outputFile, result);
        if (outputFile && outputFile !== '-') {
            console.log(colorize('Result written to ' + outputFile, 'green'));
        }
    } catch (err) {
        console.log(colorize('Error writing output: ' + err.message, 'red'));
        process.exit(1);
    }
}

main();

File: rle_decompress.rb
#!/usr/bin/env ruby
# rle_decompress.rb
# encoding: UTF-8

require 'stringio'
require 'optparse'

COLORS = {
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  reset: "\e[0m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

ESCAPE = '\\'

def decompress(text)
  return '' if text.empty?
  result = StringIO.new
  i = 0
  n = text.length
  while i < n
    ch = text[i]
    if ch == ESCAPE
      if i + 1 < n && text[i + 1] == ESCAPE
        result << ESCAPE
        i += 2
        next
      end
      raise "Unexpected end after escape" if i + 1 >= n
      repeat_char = text[i + 1]
      i += 2
      num_str = ''
      while i < n && text[i] =~ /[0-9]/
        num_str << text[i]
        i += 1
      end
      raise "Missing number after escape" if num_str.empty?
      count = num_str.to_i
      result << repeat_char * count
    else
      result << ch
      i += 1
    end
  end
  result.string
end

def read_input(filename)
  if filename.nil? || filename == '-'
    $stdin.read
  else
    File.read(filename, encoding: 'UTF-8')
  end
end

def write_output(filename, content)
  if filename.nil? || filename == '-'
    print content
  else
    File.write(filename, content, encoding: 'UTF-8')
  end
end

options = { verbose: false }
parser = OptionParser.new do |opts|
  opts.banner = "Usage: rle_decompress.rb <input> [output] [-v]"
  opts.on("-v", "--verbose", "Show statistics") { options[:verbose] = true }
end
parser.parse!

args = ARGV
if args.empty?
  puts colorize(parser.banner, :yellow)
  exit 1
end

input_file = args[0]
output_file = args[1]

begin
  data = read_input(input_file)
rescue => e
  puts colorize("Error reading input: #{e.message}", :red)
  exit 1
end

input_size = data.bytesize

begin
  result = decompress(data)
rescue => e
  puts colorize("Decompression error: #{e.message}", :red)
  exit 1
end

output_size = result.bytesize

if options[:verbose]
  ratio = input_size > 0 ? output_size.to_f / input_size : 1.0
  puts colorize("Compressed size: #{input_size} bytes", :yellow)
  puts colorize("Decompressed size: #{output_size} bytes", :yellow)
  puts colorize("Expansion ratio: #{ratio.round(2)}x", :green)
end

begin
  write_output(output_file, result)
  if output_file && output_file != '-'
    puts colorize("Result written to #{output_file}", :green)
  end
rescue => e
  puts colorize("Error writing output: #{e.message}", :red)
  exit 1
end
